File: tools/typst-preview-frontend/src/annotations.ts
/**
 * Annotation system for Typst preview
 * Provides overlay functionality for arrows, highlight boxes, and animations
 */

export interface Point {
  x: number;
  y: number;
}

export interface Bounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

export enum AnnotationType {
  ARROW = 'arrow',
  HIGHLIGHT_BOX = 'highlight-box'
}

export interface BaseAnnotation {
  id: string;
  type: AnnotationType;
  pageNumber: number;
  zIndex: number;
  opacity: number;
  visible: boolean;
  created: number;
  modified: number;
}

export interface ArrowAnnotation extends BaseAnnotation {
  type: AnnotationType.ARROW;
  start: Point;
  end: Point;
  color: string;
  thickness: number;
  arrowHeadSize: number;
  style: 'solid' | 'dashed' | 'dotted';
}

export interface HighlightBoxAnnotation extends BaseAnnotation {
  type: AnnotationType.HIGHLIGHT_BOX;
  bounds: Bounds;
  color: string;
  borderColor?: string;
  borderWidth: number;
  cornerRadius: number;
  style: 'solid' | 'dashed' | 'dotted';
}

export type Annotation = ArrowAnnotation | HighlightBoxAnnotation;

export interface AnimationKeyframe {
  time: number; // Time in milliseconds
  properties: Partial<Annotation>;
}

export interface AnimationTrack {
  annotationId: string;
  keyframes: AnimationKeyframe[];
  duration: number;
  loop: boolean;
  easing: 'linear' | 'ease-in' | 'ease-out' | 'ease-in-out';
}

export interface AnnotationDocument {
  version: string;
  documentId: string;
  annotations: Annotation[];
  animations: AnimationTrack[];
  metadata: {
    created: number;
    modified: number;
    author?: string;
    description?: string;
  };
}

export class AnnotationManager {
  private annotations: Map<string, Annotation> = new Map();
  private animations: Map<string, AnimationTrack> = new Map();
  private selectedAnnotation: string | null = null;
  private isEditMode: boolean = false;
  private currentTool: AnnotationType | null = null;
  private overlayContainer: SVGElement | null = null;
  private eventListeners: Map<string, Function[]> = new Map();

  constructor(private containerElement: HTMLElement) {
    this.setupOverlay();
    this.setupEventListeners();
  }

  private setupOverlay(): void {
    // Create SVG overlay that sits on top of the typst content
    this.overlayContainer = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    this.overlayContainer.id = 'annotation-overlay';
    this.overlayContainer.style.position = 'absolute';
    this.overlayContainer.style.top = '0';
    this.overlayContainer.style.left = '0';
    this.overlayContainer.style.width = '100%';
    this.overlayContainer.style.height = '100%';
    this.overlayContainer.style.pointerEvents = 'none';
    this.overlayContainer.style.zIndex = '1000';
    
    // Create groups for different types of annotations
    const arrowGroup = document.createElementNS('http://www.w3.org/2000/svg', 'g');
    arrowGroup.id = 'annotation-arrows';
    this.overlayContainer.appendChild(arrowGroup);
    
    const highlightGroup = document.createElementNS('http://www.w3.org/2000/svg', 'g');
    highlightGroup.id = 'annotation-highlights';
    this.overlayContainer.appendChild(highlightGroup);
    
    this.containerElement.appendChild(this.overlayContainer);
  }

  private setupEventListeners(): void {
    // Setup mouse events for creating and editing annotations
    this.containerElement.addEventListener('mousedown', this.handleMouseDown.bind(this));
    this.containerElement.addEventListener('mousemove', this.handleMouseMove.bind(this));
    this.containerElement.addEventListener('mouseup', this.handleMouseUp.bind(this));
    this.containerElement.addEventListener('keydown', this.handleKeyDown.bind(this));
  }

  private handleMouseDown(event: MouseEvent): void {
    if (!this.isEditMode || !this.currentTool) return;
    
    const point = this.getRelativePoint(event);
    this.startCreatingAnnotation(point);
  }

  private handleMouseMove(event: MouseEvent): void {
    if (!this.isEditMode) return;
    
    const point = this.getRelativePoint(event);
    this.updateCreatingAnnotation(point);
  }

  private handleMouseUp(event: MouseEvent): void {
    if (!this.isEditMode) return;
    
    const point = this.getRelativePoint(event);
    this.finishCreatingAnnotation(point);
  }

  private handleKeyDown(event: KeyboardEvent): void {
    if (event.key === 'Delete' && this.selectedAnnotation) {
      this.removeAnnotation(this.selectedAnnotation);
    }
  }

  private getRelativePoint(event: MouseEvent): Point {
    const rect = this.containerElement.getBoundingClientRect();
    return {
      x: event.clientX - rect.left,
      y: event.clientY - rect.top
    };
  }

  private startCreatingAnnotation(point: Point): void {
    // Implementation depends on current tool
    // This will be expanded in subsequent iterations
  }

  private updateCreatingAnnotation(point: Point): void {
    // Update preview of annotation being created
  }

  private finishCreatingAnnotation(point: Point): void {
    // Finalize annotation creation
  }

  public addAnnotation(annotation: Annotation): void {
    this.annotations.set(annotation.id, annotation);
    this.renderAnnotation(annotation);
    this.emit('annotationAdded', annotation);
  }

  public removeAnnotation(id: string): void {
    const annotation = this.annotations.get(id);
    if (annotation) {
      this.annotations.delete(id);
      this.removeAnnotationElement(id);
      this.emit('annotationRemoved', annotation);
    }
  }

  public updateAnnotation(id: string, updates: Partial<Annotation>): void {
    const annotation = this.annotations.get(id);
    if (annotation) {
      const updated = { ...annotation, ...updates, modified: Date.now() } as Annotation;
      this.annotations.set(id, updated);
      this.renderAnnotation(updated);
      this.emit('annotationUpdated', updated);
    }
  }

  public selectAnnotation(id: string | null): void {
    // Clear previous selection
    if (this.selectedAnnotation) {
      this.updateAnnotationSelection(this.selectedAnnotation, false);
    }
    
    this.selectedAnnotation = id;
    
    // Show new selection
    if (id) {
      this.updateAnnotationSelection(id, true);
    }
    
    this.emit('selectionChanged', id);
  }

  private updateAnnotationSelection(id: string, selected: boolean): void {
    const element = document.getElementById(`annotation-${id}`);
    if (element) {
      if (selected) {
        element.classList.add('annotation-selected');
      } else {
        element.classList.remove('annotation-selected');
      }
    }
  }

  public setEditMode(enabled: boolean): void {
    this.isEditMode = enabled;
    if (this.overlayContainer) {
      this.overlayContainer.style.pointerEvents = enabled ? 'auto' : 'none';
    }
    this.emit('editModeChanged', enabled);
  }

  public setCurrentTool(tool: AnnotationType | null): void {
    this.currentTool = tool;
    this.emit('toolChanged', tool);
  }

  private renderAnnotation(annotation: Annotation): void {
    // Remove existing element if it exists
    this.removeAnnotationElement(annotation.id);
    
    let element: SVGElement;
    
    switch (annotation.type) {
      case AnnotationType.ARROW:
        element = this.createArrowElement(annotation as ArrowAnnotation);
        break;
      case AnnotationType.HIGHLIGHT_BOX:
        element = this.createHighlightBoxElement(annotation as HighlightBoxAnnotation);
        break;
      default:
        return;
    }
    
    element.id = `annotation-${annotation.id}`;
    element.style.opacity = annotation.opacity.toString();
    element.style.visibility = annotation.visible ? 'visible' : 'hidden';
    element.style.zIndex = annotation.zIndex.toString();
    
    // Add interaction handlers
    element.style.pointerEvents = 'auto';
    element.addEventListener('click', () => this.selectAnnotation(annotation.id));
    
    const targetGroup = annotation.type === AnnotationType.ARROW ? 
      document.getElementById('annotation-arrows') :
      document.getElementById('annotation-highlights');
    
    if (targetGroup) {
      targetGroup.appendChild(element);
    }
  }

  private createArrowElement(annotation: ArrowAnnotation): SVGElement {
    const group = document.createElementNS('http://www.w3.org/2000/svg', 'g');
    
    // Create arrow line
    const line = document.createElementNS('http://www.w3.org/2000/svg', 'line');
    line.setAttribute('x1', annotation.start.x.toString());
    line.setAttribute('y1', annotation.start.y.toString());
    line.setAttribute('x2', annotation.end.x.toString());
    line.setAttribute('y2', annotation.end.y.toString());
    line.setAttribute('stroke', annotation.color);
    line.setAttribute('stroke-width', annotation.thickness.toString());
    line.setAttribute('stroke-dasharray', this.getStrokeDashArray(annotation.style));
    
    // Create arrowhead
    const arrowhead = this.createArrowhead(annotation);
    
    group.appendChild(line);
    group.appendChild(arrowhead);
    
    return group;
  }

  private createHighlightBoxElement(annotation: HighlightBoxAnnotation): SVGElement {
    const rect = document.createElementNS('http://www.w3.org/2000/svg', 'rect');
    rect.setAttribute('x', annotation.bounds.x.toString());
    rect.setAttribute('y', annotation.bounds.y.toString());
    rect.setAttribute('width', annotation.bounds.width.toString());
    rect.setAttribute('height', annotation.bounds.height.toString());
    rect.setAttribute('fill', annotation.color);
    rect.setAttribute('rx', annotation.cornerRadius.toString());
    
    if (annotation.borderColor && annotation.borderWidth > 0) {
      rect.setAttribute('stroke', annotation.borderColor);
      rect.setAttribute('stroke-width', annotation.borderWidth.toString());
      rect.setAttribute('stroke-dasharray', this.getStrokeDashArray(annotation.style));
    }
    
    return rect;
  }

  private createArrowhead(annotation: ArrowAnnotation): SVGElement {
    const { start, end, color, arrowHeadSize } = annotation;
    
    // Calculate arrow direction
    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const angle = Math.atan2(dy, dx);
    
    // Create arrowhead polygon
    const arrowhead = document.createElementNS('http://www.w3.org/2000/svg', 'polygon');
    
    const headLength = arrowHeadSize;
    const headAngle = Math.PI / 6; // 30 degrees
    
    const x1 = end.x - headLength * Math.cos(angle - headAngle);
    const y1 = end.y - headLength * Math.sin(angle - headAngle);
    const x2 = end.x - headLength * Math.cos(angle + headAngle);
    const y2 = end.y - headLength * Math.sin(angle + headAngle);
    
    const points = `${end.x},${end.y} ${x1},${y1} ${x2},${y2}`;
    arrowhead.setAttribute('points', points);
    arrowhead.setAttribute('fill', color);
    
    return arrowhead;
  }

  private getStrokeDashArray(style: string): string {
    switch (style) {
      case 'dashed':
        return '5,5';
      case 'dotted':
        return '2,2';
      default:
        return 'none';
    }
  }

  private removeAnnotationElement(id: string): void {
    const element = document.getElementById(`annotation-${id}`);
    if (element) {
      element.remove();
    }
  }

  public exportAnnotations(): AnnotationDocument {
    return {
      version: '1.0.0',
      documentId: '', // Will be set based on current document
      annotations: Array.from(this.annotations.values()),
      animations: Array.from(this.animations.values()),
      metadata: {
        created: Date.now(),
        modified: Date.now()
      }
    };
  }

  public importAnnotations(doc: AnnotationDocument): void {
    this.clearAllAnnotations();
    
    doc.annotations.forEach(annotation => {
      this.addAnnotation(annotation);
    });
    
    doc.animations.forEach(animation => {
      this.animations.set(animation.annotationId, animation);
    });
    
    this.emit('annotationsImported', doc);
  }

  public clearAllAnnotations(): void {
    this.annotations.clear();
    this.animations.clear();
    
    const arrowGroup = document.getElementById('annotation-arrows');
    const highlightGroup = document.getElementById('annotation-highlights');
    
    if (arrowGroup) arrowGroup.innerHTML = '';
    if (highlightGroup) highlightGroup.innerHTML = '';
  }

  // Event system
  public on(event: string, callback: Function): void {
    if (!this.eventListeners.has(event)) {
      this.eventListeners.set(event, []);
    }
    this.eventListeners.get(event)!.push(callback);
  }

  public off(event: string, callback: Function): void {
    const listeners = this.eventListeners.get(event);
    if (listeners) {
      const index = listeners.indexOf(callback);
      if (index > -1) {
        listeners.splice(index, 1);
      }
    }
  }

  private emit(event: string, data?: any): void {
    const listeners = this.eventListeners.get(event);
    if (listeners) {
      listeners.forEach(callback => callback(data));
    }
  }

  public dispose(): void {
    this.clearAllAnnotations();
    this.eventListeners.clear();
    
    if (this.overlayContainer) {
      this.overlayContainer.remove();
    }
  }
}

File: crates/typst-preview/src/annotations.rs
use std::collections::HashMap;
use std::sync::Arc;
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

/// Point coordinate for annotation positioning
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Point {
    pub x: f64,
    pub y: f64,
}

/// Bounding rectangle for annotation positioning
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Bounds {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

/// Types of annotations supported
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AnnotationType {
    Arrow,
    HighlightBox,
}

/// Base annotation properties
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BaseAnnotation {
    pub id: String,
    pub annotation_type: AnnotationType,
    pub page_number: u32,
    pub z_index: i32,
    pub opacity: f64,
    pub visible: bool,
    pub created: u64,
    pub modified: u64,
}

/// Arrow annotation specific properties
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArrowAnnotation {
    #[serde(flatten)]
    pub base: BaseAnnotation,
    pub start: Point,
    pub end: Point,
    pub color: String,
    pub thickness: f64,
    pub arrow_head_size: f64,
    pub style: String, // solid, dashed, dotted
}

/// Highlight box annotation specific properties
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HighlightBoxAnnotation {
    #[serde(flatten)]
    pub base: BaseAnnotation,
    pub bounds: Bounds,
    pub color: String,
    pub border_color: Option<String>,
    pub border_width: f64,
    pub corner_radius: f64,
    pub style: String, // solid, dashed, dotted
}

/// Union type for all annotation types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "annotation_type", rename_all = "kebab-case")]
pub enum Annotation {
    Arrow(ArrowAnnotation),
    HighlightBox(HighlightBoxAnnotation),
}

impl Annotation {
    pub fn id(&self) -> &str {
        match self {
            Annotation::Arrow(a) => &a.base.id,
            Annotation::HighlightBox(a) => &a.base.id,
        }
    }

    pub fn page_number(&self) -> u32 {
        match self {
            Annotation::Arrow(a) => a.base.page_number,
            Annotation::HighlightBox(a) => a.base.page_number,
        }
    }

    pub fn set_modified(&mut self, timestamp: u64) {
        match self {
            Annotation::Arrow(a) => a.base.modified = timestamp,
            Annotation::HighlightBox(a) => a.base.modified = timestamp,
        }
    }
}

/// Animation keyframe for annotation properties
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnimationKeyframe {
    pub time: f64, // Time in milliseconds
    pub properties: serde_json::Value, // Flexible property updates
}

/// Animation track for a specific annotation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnimationTrack {
    pub annotation_id: String,
    pub keyframes: Vec<AnimationKeyframe>,
    pub duration: f64,
    pub loop_animation: bool,
    pub easing: String, // linear, ease-in, ease-out, ease-in-out
}

/// Complete annotation document with metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnnotationDocument {
    pub version: String,
    pub document_id: String,
    pub annotations: Vec<Annotation>,
    pub animations: Vec<AnimationTrack>,
    pub metadata: AnnotationMetadata,
}

/// Metadata for annotation documents
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnnotationMetadata {
    pub created: u64,
    pub modified: u64,
    pub author: Option<String>,
    pub description: Option<String>,
}

/// Annotation manager for storing and retrieving annotations
pub struct AnnotationManager {
    /// Map from document path to annotation document
    annotations_by_document: Arc<RwLock<HashMap<String, AnnotationDocument>>>,
    /// Map from annotation ID to document path for quick lookups
    annotation_index: Arc<RwLock<HashMap<String, String>>>,
}

impl AnnotationManager {
    pub fn new() -> Self {
        Self {
            annotations_by_document: Arc::new(RwLock::new(HashMap::new())),
            annotation_index: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Get all annotations for a specific document
    pub async fn get_annotations(&self, document_path: &str) -> Option<AnnotationDocument> {
        let annotations = self.annotations_by_document.read().await;
        annotations.get(document_path).cloned()
    }

    /// Add or update an annotation document
    pub async fn set_annotations(&self, document_path: String, mut doc: AnnotationDocument) {
        doc.metadata.modified = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        // Update annotation index
        let mut index = self.annotation_index.write().await;
        for annotation in &doc.annotations {
            index.insert(annotation.id().to_string(), document_path.clone());
        }
        drop(index);

        // Store the document
        let mut annotations = self.annotations_by_document.write().await;
        annotations.insert(document_path, doc);
    }

    /// Add a single annotation to a document
    pub async fn add_annotation(&self, document_path: &str, annotation: Annotation) -> Result<(), String> {
        let mut annotations = self.annotations_by_document.write().await;
        
        let doc = annotations.get_mut(document_path).ok_or("Document not found")?;
        
        // Check if annotation already exists
        if doc.annotations.iter().any(|a| a.id() == annotation.id()) {
            return Err("Annotation with this ID already exists".to_string());
        }

        doc.annotations.push(annotation.clone());
        doc.metadata.modified = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        // Update index
        drop(annotations);
        let mut index = self.annotation_index.write().await;
        index.insert(annotation.id().to_string(), document_path.to_string());

        Ok(())
    }

    /// Update an existing annotation
    pub async fn update_annotation(&self, annotation_id: &str, mut annotation: Annotation) -> Result<(), String> {
        let index = self.annotation_index.read().await;
        let document_path = index.get(annotation_id).ok_or("Annotation not found")?.clone();
        drop(index);

        let mut annotations = self.annotations_by_document.write().await;
        let doc = annotations.get_mut(&document_path).ok_or("Document not found")?;

        let annotation_pos = doc.annotations.iter().position(|a| a.id() == annotation_id)
            .ok_or("Annotation not found in document")?;

        annotation.set_modified(std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64);

        doc.annotations[annotation_pos] = annotation;
        doc.metadata.modified = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        Ok(())
    }

    /// Remove an annotation
    pub async fn remove_annotation(&self, annotation_id: &str) -> Result<(), String> {
        let index = self.annotation_index.read().await;
        let document_path = index.get(annotation_id).ok_or("Annotation not found")?.clone();
        drop(index);

        let mut annotations = self.annotations_by_document.write().await;
        let doc = annotations.get_mut(&document_path).ok_or("Document not found")?;

        let annotation_pos = doc.annotations.iter().position(|a| a.id() == annotation_id)
            .ok_or("Annotation not found in document")?;

        doc.annotations.remove(annotation_pos);
        doc.metadata.modified = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        // Remove from index
        drop(annotations);
        let mut index = self.annotation_index.write().await;
        index.remove(annotation_id);

        Ok(())
    }

    /// Clear all annotations for a document
    pub async fn clear_annotations(&self, document_path: &str) -> Result<(), String> {
        let mut annotations = self.annotations_by_document.write().await;
        let doc = annotations.get_mut(document_path).ok_or("Document not found")?;

        // Remove all annotations from index
        let mut index = self.annotation_index.write().await;
        for annotation in &doc.annotations {
            index.remove(annotation.id());
        }
        drop(index);

        doc.annotations.clear();
        doc.metadata.modified = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        Ok(())
    }

    /// Create a new empty annotation document for a path
    pub async fn create_document(&self, document_path: String) -> AnnotationDocument {
        let timestamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        let doc = AnnotationDocument {
            version: "1.0.0".to_string(),
            document_id: document_path.clone(),
            annotations: Vec::new(),
            animations: Vec::new(),
            metadata: AnnotationMetadata {
                created: timestamp,
                modified: timestamp,
                author: None,
                description: None,
            },
        };

        let mut annotations = self.annotations_by_document.write().await;
        annotations.insert(document_path, doc.clone());

        doc
    }

    /// Export annotations to JSON
    pub async fn export_annotations(&self, document_path: &str) -> Result<String, String> {
        let annotations = self.annotations_by_document.read().await;
        let doc = annotations.get(document_path).ok_or("Document not found")?;
        serde_json::to_string_pretty(doc).map_err(|e| e.to_string())
    }

    /// Import annotations from JSON
    pub async fn import_annotations(&self, document_path: String, json_data: &str) -> Result<(), String> {
        let doc: AnnotationDocument = serde_json::from_str(json_data)
            .map_err(|e| format!("Failed to parse annotation data: {}", e))?;

        self.set_annotations(document_path, doc).await;
        Ok(())
    }
}

impl Default for AnnotationManager {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_annotation_manager() {
        let manager = AnnotationManager::new();
        let doc_path = "/test/document.typ".to_string();

        // Create a new document
        let doc = manager.create_document(doc_path.clone()).await;
        assert_eq!(doc.annotations.len(), 0);

        // Create an arrow annotation
        let arrow = Annotation::Arrow(ArrowAnnotation {
            base: BaseAnnotation {
                id: "arrow1".to_string(),
                annotation_type: AnnotationType::Arrow,
                page_number: 1,
                z_index: 1,
                opacity: 1.0,
                visible: true,
                created: 0,
                modified: 0,
            },
            start: Point { x: 10.0, y: 10.0 },
            end: Point { x: 50.0, y: 50.0 },
            color: "#ff0000".to_string(),
            thickness: 2.0,
            arrow_head_size: 10.0,
            style: "solid".to_string(),
        });

        // Add annotation
        manager.add_annotation(&doc_path, arrow).await.unwrap();

        // Retrieve and verify
        let retrieved_doc = manager.get_annotations(&doc_path).await.unwrap();
        assert_eq!(retrieved_doc.annotations.len(), 1);
        assert_eq!(retrieved_doc.annotations[0].id(), "arrow1");

        // Test export/import
        let exported = manager.export_annotations(&doc_path).await.unwrap();
        assert!(exported.contains("arrow1"));

        // Clear and import
        manager.clear_annotations(&doc_path).await.unwrap();
        let empty_doc = manager.get_annotations(&doc_path).await.unwrap();
        assert_eq!(empty_doc.annotations.len(), 0);

        manager.import_annotations(doc_path.clone(), &exported).await.unwrap();
        let imported_doc = manager.get_annotations(&doc_path).await.unwrap();
        assert_eq!(imported_doc.annotations.len(), 1);
    }
}